lib/matplotlib/axes/_axes.py

```diff
@@ -6207,8 +6216,9 @@ def pcolormesh(self, *args, alpha=None, norm=None, cmap=None, vmin=None,
         X, Y, C, shading = self._pcolorargs('pcolormesh', *args,
                                             shading=shading, kwargs=kwargs)
         coords = np.stack([X, Y], axis=-1)
-        # convert to one dimensional array
-        C = C.ravel()
+        # convert to one dimensional array, except for 3D RGB(A) arrays
+        if C.ndim != 3:
+            C = C.ravel()
```

Comment:
Why do 3-D arrays not need to be flattened at all?

Reply:
or conversely, why are we flattening 2-D arrays?

Reply:
Yes, I am not a fan of the flattening either... Allowing 2-D arrays is pretty new (only the last couple of releases), so everyone always used to have to flatten their arrays before getting/setting arrays. Unfortunately, that is baked into a few of our tests as expecting to pass a 2-D array in, and then calling `mesh = plt.pcolormesh(z2d); mesh.get_array() == z2d.ravel()`. I don't think we should be messing with people's arrays if they pass them in as 2D, but that may be too controversial to change at this point too...

Reply:
> Why do 3-D arrays not need to be flattened at all?
Specifically for 3-D, we can not flatten them because the image RGB(A) color handling code will only handle the 3-D shapes of `(M, N, [3 or 4])`